src/sentry/discover/dataset_split.py

```diff
@@ -0,0 +1,386 @@
+    "span_op_breakdowns[ops.browser]",
+    "span_op_breakdowns[ops.http]",
+    "span_op_breakdowns[ops.db]",
+    "span_op_breakdowns[ops.resource]",
+    "span_op_breakdowns[ops.ui]",
+]
+
+ERROR_ONLY_FIELDS = [
+    "location",
+    "exception_stacks.type",
+    "exception_stacks.value",
+    "exception_stacks.mechanism_type",
+    "exception_stacks.mechanism_handled",
+    "received",
+    "exception_main_thread",
+    "exception_frames.abs_path",
+    "exception_frames.colno",
+    "exception_frames.filename",
+    "exception_frames.function",
+    "exception_frames.in_app",
+    "exception_frames.lineno",
+    "exception_frames.module",
+    "exception_frames.package",
+    "exception_frames.stack_level",
+]
+
+
+def save_split_decision_for_query(
+    saved_query: DiscoverSavedQuery,
+    split_decision: int | None,
+    dataset_source: DatasetSourcesTypes | None,
+):
+    if split_decision is not None:
+        saved_query.dataset = split_decision
+    if dataset_source is not None:
+        saved_query.dataset_source = dataset_source.value
+
+    saved_query.save()
+
+
+def check_function_parameter_matches_dataset(
+    function: Function | CurriedFunction,
+    dataset: Dataset,
+) -> bool:
+    fields = TRANSACTION_ONLY_FIELDS if dataset == Dataset.Transactions else ERROR_ONLY_FIELDS
+    for parameter in function.parameters:
+        if isinstance(parameter, Column) and parameter.name in fields:
+            return True
+
+    return False
+
+
+def check_aliased_expression_matches_dataset(
+    aliased_exp: AliasedExpression,
+    dataset: Dataset,
+) -> bool:
+    col = aliased_exp.exp.name
+    fields = TRANSACTION_ONLY_FIELDS if dataset == Dataset.Transactions else ERROR_ONLY_FIELDS
+    if col in fields:
+        return True
+
+    return False
+
+
+def check_column_matches_dataset(
+    column: Column,
+    dataset: Dataset,
+) -> bool:
+    col = column.name
+    fields = TRANSACTION_ONLY_FIELDS if dataset == Dataset.Transactions else ERROR_ONLY_FIELDS
+    if col in fields:
+        return True
+
+    return False
+
+
+def check_event_type_condition(cond: Condition, dataset: Dataset):
+    op = cond.op
+    rhs = cond.rhs
+    if dataset == Dataset.Events and (
+        (op == Op.EQ and rhs == "error") or (op == Op.NEQ and rhs == "transaction")
+    ):
+        return True
+
+    if dataset == Dataset.Transactions and op == Op.EQ and rhs == "transaction":
+        return True
+
+    return False
+
+
+def check_condition_matches_dataset(
+    cond: Condition,
+    dataset: Dataset,
+) -> bool:
+    lhs = cond.lhs
+    if isinstance(lhs, Column):
+        if lhs.name == "type":
+            return check_event_type_condition(cond, dataset)
+
+        return check_column_matches_dataset(lhs, dataset)
+
+    if isinstance(lhs, Function) or isinstance(lhs, CurriedFunction):
+        return check_function_parameter_matches_dataset(lhs, dataset)
+
+    return False
+
+
+def check_top_level_conditions_match_dataset(
```

Comment:
The terminology you used here, `top_level_conditions`, are there conditions that are not top level?

Reply:
We basically don't bother checking nested conditions like (foo AND (bar OR baz)) will only check foo condition